pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/rebalance/DefaultRebalanceSegmentStrategy.java

```diff
@@ -0,0 +1,333 @@
+
+  private HelixManager _helixManager;
+  private HelixAdmin _helixAdmin;
+  private String _helixClusterName;
+  private ZkHelixPropertyStore<ZNRecord> _propertyStore;
+
+  public DefaultRebalanceSegmentStrategy(HelixManager helixManager) {
+    super(helixManager);
+    _helixManager = helixManager;
+    _helixAdmin = helixManager.getClusterManagmentTool();
+    _helixClusterName = helixManager.getClusterName();
+    _propertyStore = helixManager.getHelixPropertyStore();
+  }
+
+  /**
+   * Rebalance partition assignment only for realtime tables, and if includeConsuming=true
+   * @param idealState old ideal state
+   * @param tableConfig table config of table tor rebalance
+   * @param rebalanceUserConfig custom user configs for specific rebalance strategies
+   * @return
+   */
+  @Override
+  public PartitionAssignment rebalancePartitionAssignment(IdealState idealState, TableConfig tableConfig,
+      Configuration rebalanceUserConfig) {
+    PartitionAssignment newPartitionAssignment = new PartitionAssignment(tableConfig.getTableName());
+
+    if (tableConfig.getTableType().equals(CommonConstants.Helix.TableType.REALTIME)) {
+      LOGGER.info("Rebalancing stream partition assignment for table {}", tableConfig.getTableName());
+
+      boolean includeConsuming =
+          rebalanceUserConfig.getBoolean(RebalanceUserConfigConstants.INCLUDE_CONSUMING, DEFAULT_INCLUDE_CONSUMING);
+      if (includeConsuming) {
+
+        String tableNameWithType = tableConfig.getTableName();
+
+        StreamPartitionAssignmentGenerator streamPartitionAssignmentGenerator =
+            new StreamPartitionAssignmentGenerator(_propertyStore);
+        PartitionAssignment streamPartitionAssignment =
+            streamPartitionAssignmentGenerator.getStreamPartitionAssignment(tableNameWithType);
+        int numPartitions = streamPartitionAssignment.getNumPartitions();
+
+        RealtimeTagConfig realtimeTagConfig = new RealtimeTagConfig(tableConfig, _helixManager);
+        List<String> consumingInstances =
+            _helixAdmin.getInstancesInClusterWithTag(_helixClusterName, realtimeTagConfig.getConsumingServerTag());
+
+        Map<String, PartitionAssignment> tableNameToStreamPartitionAssignmentMap =
+            streamPartitionAssignmentGenerator.generatePartitionAssignment(tableConfig, numPartitions,
+                consumingInstances, Lists.newArrayList(tableNameWithType));
+        newPartitionAssignment = tableNameToStreamPartitionAssignmentMap.get(tableNameWithType);
+
+        boolean dryRun = rebalanceUserConfig.getBoolean(RebalanceUserConfigConstants.DRYRUN, DEFAULT_DRY_RUN);
+        if (!dryRun) {
+          LOGGER.info("Updating stream partition assignment for table {}", tableNameWithType);
+          streamPartitionAssignmentGenerator.writeStreamPartitionAssignment(tableNameToStreamPartitionAssignmentMap);
+        } else {
+          LOGGER.info("Dry run. Skip writing stream partition assignment to property store");
```

Comment:
we should find a way to include new partition assignment in the output of dryrun

Reply:
The api has already been modified to send back partition assignment along with the ideal state